feat: Add multi_match support for fuzzy

smessie opened the pull request:
This PR adds support to match multiple fields (or all with `_all`) for the fuzzy flag

madnificent commented:
@nvdk Can you review and merge if okay?  Thank you :-)

nvdk commented on framework/elastic_query_builder.rb:
```diff
@@ -143,7 +143,12 @@ def construct_es_query_term filter_key, value
       }
-    when "term", "fuzzy", "prefix", "wildcard", "regexp"
+    when "fuzzy"
+      multi_match_fields = fields == ["_all"] ? nil : fields
```
I would recommend setting to '*.*' instead of nil if we're matching all, making that explicit instead of relying on the elastic default for no fields specified. That may avoid some confusion if `index.query.default_field` is set to something different

smessie replied:
Based myself on the case above it where it is also set to nil, but will change it to `*.*`

nvdk commented on framework/elastic_query_builder.rb:
```diff
@@ -143,7 +143,12 @@ def construct_es_query_term filter_key, value
+    when "fuzzy"
+      multi_match_fields = fields == ["_all"] ? nil : fields
+      {
+        multi_match: { query: value, fields: multi_match_fields, fuzziness: "AUTO" }.compact
```
I think the compact can be dropped if we implement the above suggestion

nvdk replied:
perhaps better to keep it, if it's allowed to have fields unspecified in the api (I'm unsure)

nvdk commented:
> This PR adds support to match multiple fields (or all with `_all`) for the fuzzy flag

wasn't this already possible with fuzzy_match?

nvdk commented:
I think the es query builder in general should be looked at, we're now adding this for fuzzy but shouldn't the same be possible for a regular query?

> > This PR adds support to match multiple fields (or all with `_all`) for the fuzzy flag
> 
> wasn't this already possible with fuzzy_match?

t wasn't, but I don't really see the difference between fuzzy_match and fuzzy with the changes. perhaps there wasn't a meaningful one to begin with. 

I do feel like this addresses just a small part, should we create an issue to tackle this for the other types as well? I think the same logic applies to phrase, fuzzy_phrase, perhaps others

nvdk reviewed (changes requested):
So in summary:
- by setting fuzziness to auto for "fuzzy" we've eliminated the difference between fuzzy and fuzzy_match. This at least should be corrected.
- for `_all` I recommend setting fields to `*.*` explicitly instead of relying on the default fields of the index
- I think we should allow multi matching also in other places where applicable, for example fuzzy_match and phrase_match
- the README should be updated to reflect the changes, see https://github.com/mu-semtech/mu-search/blob/master/README.md#custom-queries
-

smessie commented:
> So in summary:
> 
> * by setting fuzziness to auto for "fuzzy" we've eliminated the difference between fuzzy and fuzzy_match. This at least should be corrected.
> * for `_all` I recommend setting fields to `*.*` explicitly instead of relying on the default fields of the index
> * I think we should allow multi matching also in other places where applicable, for example fuzzy_match and phrase_match
> * the README should be updated to reflect the changes, see https://github.com/mu-semtech/mu-search/blob/master/README.md#custom-queries

I've removed `fuzzy_match` in favor of `fuzzy` which now thus has fuzziness set to AUTO and allows for matching multiple fields. I've also updated the README.
I've changed `nil` to `*.*`.
Should we just open an issue for the others or do you want this to get this fixed in this PR as well? Is matching multiple fields applicable to every term? I'm not very informed of the behavior of every term, I just added the missing one that I needed and thus was able to test live.

smessie commented:
Okay @nvdk so, when I was using this last version I encountered an exception: 
`{"type":"illegal_argument_exception","reason":"Can only use fuzzy queries on keyword and text fields - not on [price.price] which is of type [double]"}`
This is because of replacing `nil` with `*.*`, seems like `nil` will under the hood just match all possible fields, while `*.*` matches all (also conflicting) fields.
I would thus say that it's better to change it to `nil` again?

nvdk commented:
ah that's interesting an probably the reason we went with nil before, could you indeed revert to using nil and add a comment to that effect so we know later on? Tx!

Wrt to other types, matching multiple fields is only possible for fuzzy, phrase and phrase_prefix I think. I'd have to look at the difference between phrase_match and phrase in musearch. might be similar to the fuzzy vs fuzzy_match where it makes sense to drop phrase_match.

smessie commented:
Done!

nvdk commented on README.md:
````diff
@@ -214,7 +214,7 @@ This guide assumes you have already integrated mu-search in your application and
 For indexing files mu-search requires a Tika server to extract the content. Add the `tika` service next to the `search` and `elasticsearch` services in `docker-compose.yml`:
 ```yml
-services:
+services:m
````
there's a typo here :)

nvdk commented on README.md:
should add fuzzy to the list here

smessie replied:
Fuzzy is currently under "Term-level queries", do you want it to be moved under "Custom queries" or do you rather prefer it to be included in both parts?

nvdk replied:
mostly in the list " Currently searching on multiple fields is only supported for the following flag: "

smessie replied:
Done :)

nvdk commented:
@smessie I added some remarks on the readme, if those are fixed this can be merged!
